File: 937-online-stock-span/solution.ts
class StockSpanner {
  private stack: [number, number][] = [];
  private idx = 0;

  next(price: number): number {
    while (price >= this.stack[this.stack.length - 1]?.[1]) this.stack.pop();
    const prev = this.stack[this.stack.length - 1]?.[0] ?? -1;
    this.stack.push([this.idx, price]);
    return this.idx++ - prev;
  }
}

/**
 * Your StockSpanner object will be instantiated and called as such:
 * var obj = new StockSpanner()
 * var param_1 = obj.next(price)
 */

File: 937-online-stock-span/solution.cpp
class StockSpanner {
public:
    stack<pair<int, vector<int>>> sk; // [price, idx]
    int idx = 0;
    
    StockSpanner() {}
    
    int next(int price) {
        while(!sk.empty() && sk.top().first < price) sk.pop();
        int ans;
        if(sk.empty()) ans = idx + 1;
        else if(sk.top().first == price) {
            if(sk.size() == 1) ans = idx + 1;
            else {
                auto top = sk.top();
                sk.pop();
                ans = idx - sk.top().second.back();
                sk.push(top);
            }
        }
        else ans = idx - sk.top().second.back();
        if(sk.empty() || sk.top().first > price) sk.push({price, {idx++}});
        else sk.top().second.push_back(idx++);
        return ans;
    }
};

/**
 * Your StockSpanner object will be instantiated and called as such:
 * StockSpanner* obj = new StockSpanner();
 * int param_1 = obj->next(price);
 */
